### PR TITLE
feat: add C implementation for `math/base/special/max`

### LINKED ISSUE #1433: [RFC]: Add C implementation for `@stdlib/math/base/special/max`
### Description

This RFC proposes adding native C implementation for [`@stdlib/math/base/special/max`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/max/).

```c
double stdlib_base_max( const double x, const double y );
```

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Shashankss1205**
Resolves #1433  .

## Description
This pull request adds native C implementation for
@stdlib/math/base/special/max for issue #1433 .
> What is the purpose of this pull request?
Native C implementation of max function
This pull request:

-   a
-   b
-   c

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #
-   fixes #

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/max/README.md [hunk 1]
```diff
@@ -103,6 +103,88 @@ for ( i = 0; i < 100; i++ ) {
...
+<!-- /.usage -->
+<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+<section class="notes">
+</section>
+<!-- /.notes -->
+<!-- C API usage examples. -->
+<section class="examples">
+### Examples
+```c
+#include "stdlib/math/base/special/max.h"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/README.md]: Incorrect spacing here

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile [hunk 2]
```diff
@@ -1,107 +0,0 @@
-#/
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile]: Don't delete this file, revert this commit

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/native/benchmark.c [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /**
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/c/native/benchmark.c:1]: Keep both files `c/benchmark.c` and `c/native/benchmark.c`

File: lib/node_modules/@stdlib/math/base/special/max/manifest.json [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        "libraries": [],
+        "libpath": [],
+        "dependencies": [
+          "@stdlib/math/base/assert/is-nan",
+          "@stdlib/math/base/assert/is-negative-zero",
+          "@stdlib/constants/float64/ninf"
+        ]
+      }
+    ]
+  }
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/manifest.json]: incorrect indentation

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Requested changes, once done this PR can be reviewed again.

File: lib/node_modules/@stdlib/math/base/special/max/README.md (hunk 1 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/README.md]: Changes committed

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile (hunk 2 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/native/benchmark.c (hunk 3 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/benchmark/c/native/benchmark.c:1]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/manifest.json (hunk 4 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/manifest.json]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/README.md [hunk 5]
```diff
@@ -103,6 +103,99 @@ for ( i = 0; i < 100; i++ ) {
...
+</section>
+
+<!-- /.notes -->
+
+<!-- C API usage examples. -->
+
+<section class="examples">
+    
+
+### Examples
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/README.md]:
```suggestion
<section class="examples">

### Examples
```

File: lib/node_modules/@stdlib/math/base/special/max/README.md [hunk 6]
```diff
@@ -103,6 +103,99 @@ for ( i = 0; i < 100; i++ ) {
...
+
+```c
+#include "stdlib/math/base/special/max.h"
+#include <stdlib.h>
+#include <stdio.h>
+int main( void ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/README.md]:
```suggestion
#include <stdio.h>

int main( void ) {
```

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile [hunk 7]
```diff
@@ -1,7 +1,7 @@
 #/
 # @license Apache-2.0
 #
-# Copyright (c) 2018 The Stdlib Authors.
+# Copyright (c) 2024 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile]:
```suggestion
# Copyright (c) 2018 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile [hunk 8]
```diff
@@ -104,4 +104,4 @@ run: $(c_targets)
 clean:
 	$(QUIET) -rm -f *.o *.out
 
-.PHONY: clean
+.PHONY: clean
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile]: No changes needed in this file.

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c [hunk 9]
```diff
@@ -134,4 +136,4 @@ int main( void ) {
 		printf( "ok %d benchmark finished\n", i+1 );
 	}
 	print_summary( REPEATS, REPEATS );
-}
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c]: we need no change in this file as well

File: lib/node_modules/@stdlib/math/base/special/max/src/addone.c [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/max.h"
+#include "stdlib/math/base/napi/binary.h"
+
+STDLIB_MATH_BASE_NAPI_MODULE_DD_D( stdlib_base_max )
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/src/addone.c]: incorrect file name

File: lib/node_modules/@stdlib/math/base/special/max/src/main.c [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+*/
+double stdlib_base_max( const double x, const double y ) {
+    if ( stdlib_base_is_nan( x ) || stdlib_base_is_nan( y ) ) {
+        return 0.0 / 0.0; // NaN
+    }
+    if ( x == STDLIB_CONSTANT_FLOAT64_PINF || y == STDLIB_CONSTANT_FLOAT64_PINF ) {
+        return STDLIB_CONSTANT_FLOAT64_PINF;
+    }
+    if( x == y && x == 0.0 ) {
+        if ( stdlib_base_is_positive_zero ( x ) ) {
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/src/main.c]:
```suggestion
        if ( stdlib_base_is_positive_zero( x ) ) {
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Few more changes and this be good to merge.

File: lib/node_modules/@stdlib/math/base/special/max/README.md (hunk 5 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/README.md]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/README.md (hunk 6 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/README.md]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile (hunk 7 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/benchmark/c/Makefile]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c (hunk 9 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/src/addone.c (hunk 10 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/src/addone.c]: Resolved

File: lib/node_modules/@stdlib/math/base/special/max/src/main.c (hunk 11 above)
Comment (Shashankss1205) [reply lib/node_modules/@stdlib/math/base/special/max/src/main.c]: Resolved

**Comment by Shashankss1205** [comment]
please close the PR if the issue has been resolved @Pranavchiku .

File: lib/node_modules/@stdlib/math/base/special/max/README.md [hunk 12]
```diff
@@ -2,7 +2,7 @@
 
 @license Apache-2.0
 
-Copyright (c) 2018 The Stdlib Authors.
+Copyright (c) 2024 The Stdlib Authors.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/README.md]:
```suggestion
Copyright (c) 2018 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/benchmark.native.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+			b.fail( 'should not return NaN' );
+		}
+	}
+	b.toc();
+	if ( isnan( z ) ) {
+		b.fail( 'should not return NaN' );
+	}
+	b.pass( 'benchmark finished' );
+	b.end();
+});
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/benchmark/benchmark.native.js]: add new line at end of file

File: lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c (hunk 9 above)
Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/max/benchmark/c/benchmark.c]: I don't think this is done yet, I can see the changes.

File: lib/node_modules/@stdlib/math/base/special/max/binding.gyp [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+          'action': [
+            'cp',
+            '<(PRODUCT_DIR)/<(addon_target_name).node',
+            '<(addon_output_dir)/<(addon_target_name).node',
+          ],
+        },
+      ], # end actions
+    }, # end target copy_addon
+  ], # end targets
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/binding.gyp]: newline here

File: lib/node_modules/@stdlib/math/base/special/max/examples/c/Makefile [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+#/
+# Removes generated files.
+#
+# @example
+# make clean
+#/
+clean:
+	$(QUIET) -rm -f *.o *.out
+
+.PHONY: clean
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/examples/c/Makefile]: new line here

File: lib/node_modules/@stdlib/math/base/special/max/examples/c/example.c [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    double v;
+    int i;
+
+    for ( i = 0; i < 100; i++ ) {
+        x = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        y = ( ( (double)rand() / (double)RAND_MAX ) * 200.0 ) - 100.0;
+        v = stdlib_base_max( x, y );
+        printf( "x: %lf, y: %lf, max(x, y): %lf\n", x, y, v );
+    }
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/examples/c/example.c]: newline here

File: lib/node_modules/@stdlib/math/base/special/max/include.gypi [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    'libraries': [
+      '<!@(node -e "var arr = require(\'@stdlib/utils/library-manifest\')(\'./manifest.json\',{},{\'basedir\':process.cwd(),\'paths\':\'posix\'}).libraries; for ( var i = 0; i < arr.length; i++ ) { console.log( arr[ i ] ); }")',
+    ],
+
+    # Library directories:
+    'library_dirs': [
+      '<!@(node -e "var arr = require(\'@stdlib/utils/library-manifest\')(\'./manifest.json\',{},{\'basedir\':process.cwd(),\'paths\':\'posix\'}).libpath; for ( var i = 0; i < arr.length; i++ ) { console.log( arr[ i ] ); }")',
+    ],
+  }, # end variables
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/include.gypi]: newline here

File: lib/node_modules/@stdlib/math/base/special/max/include/stdlib/math/base/special/max.h [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+/**
+* Returns the maximum value.
+*/
+double stdlib_base_max( const double x, const double y );
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif // !STDLIB_MATH_BASE_SPECIAL_MAX_H
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/include/stdlib/math/base/special/max.h]: newline here

File: lib/node_modules/@stdlib/math/base/special/max/lib/native.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+* // returns NaN
+*/
+function max( x, y ) {
+	return addon( x, y );
+}
+
+
+// EXPORTS //
+
+module.exports = max;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/lib/native.js]: newline

File: lib/node_modules/@stdlib/math/base/special/max/manifest.json [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/math/base/assert/is-negative-zero",
+        "@stdlib/constants/float64/ninf"
+      ]
+    }
+  ]
+}
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/manifest.json]: here also

File: lib/node_modules/@stdlib/math/base/special/max/src/Makefile [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+#/
+# Removes generated files.
+#
+# @example
+# make clean
+#/
+clean: clean-addon
+
+.PHONY: clean
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/max/src/Makefile]: newline

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Almost done, minor changes and this will be ready.

**Comment by Shashankss1205** [comment]
Please take a look at this once @Pranavchiku . Sorry for the extraneous commits, I will try to keep them clean from the next PR's

**Comment by Shashankss1205** [comment]
Please take a look at this once @Pranavchiku @Planeshifter. 

Sorry for the extraneous commits, I will try to keep them clean from the next PR's I am just a beginner in open source.

File: lib/node_modules/@stdlib/math/base/special/max/src/main.c [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+* // returns 0.0
+*/
+double stdlib_base_max( const double x, const double y ) {
+    if ( stdlib_base_is_nan( x ) || stdlib_base_is_nan( y ) ) {
+        return 0.0 / 0.0; // NaN
+    }
+    if ( x == STDLIB_CONSTANT_FLOAT64_PINF || y == STDLIB_CONSTANT_FLOAT64_PINF ) {
+        return STDLIB_CONSTANT_FLOAT64_PINF;
+    }
+    if( x == y && x == 0.0 ) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/max/src/main.c]:
```suggestion
    if ( x == y && x == 0.0 ) {
```

File: lib/node_modules/@stdlib/math/base/special/max/src/main.c [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+* limitations under the License.
+*/
+
+#include "stdlib/math/base/special/max.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/assert/is_positive_zero.h"
+#include "stdlib/constants/float64/pinf.h"
+
+/**
+* Return the maximum value.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/max/src/main.c]:
```suggestion
* Returns the maximum value.
```

**Review by Planeshifter — APPROVED** [review]
LGTM! Will merge once CI checks have cleared.

Thank you!

**Pull request merged**